cores/arduino/HardwareTimer.cpp

```diff
@@ -680,10 +843,21 @@ void HardwareTimer::detachInterrupt()
   */
 void HardwareTimer::attachInterrupt(uint32_t channel, void (*callback)(HardwareTimer *))
 {
+  int interrupt = getIT(channel);
+  if (interrupt == -1) {
+    Error_Handler();
+  }
+
   if ((channel == 0) || (channel > (TIMER_CHANNELS + 1))) {
     Error_Handler();  // only channel 1..4 have an interrupt
   }
 
+  if (callback != NULL) {
+    // Clear flag before enabling IT
+    __HAL_TIM_CLEAR_FLAG(&(_timerObj.handle), interrupt);
```

Comment:
I am not entirely sure about the timing details on STM32, but I can imagine that the interrupt arrives at the same time as the clear, and thus is cleared before it can run.

Another case is when code is called when interrupts are temporarily disabled (e.g. using the global interrupt mask, or when running from an ISR with higher priority). In that case, the flag might be set before the clear, but the interrupt cannot run yet, so the clear loses one interrupt.

> And Clear is necessary for example in the following use case:
The timer is already running with interrupts disabled. After a while, Flags Update or Compare will be set. Then if we enable interrupt we must clear the flag to discard past interrupts and take into account only new ones.

Yeah, I agree on that. Hence my suggestion to only clear if the interrupt is not enabled yet, which I think should cover the case you describe? If the interrupt is already enabled, then the clear is not needed, assuming that any interrupts that are still pending should be handled by the new callback rather than be lost.

Reply:
Right, I agree with your proposal. I implement it in https://github.com/stm32duino/Arduino_Core_STM32/pull/852